Introduce a use case-driven section on "managing transitive dependencies" in the user guide

bmuschko opened the pull request:
### Context

This is the initial draft of a use case-driven section on "managing transitive dependencies" for the user guide.

- Almost completely deletes the existing content.
- New content was added based on the proposed outline.
    - Excluding transitive module dependencies
    - Enforcing a particular dependency version
    - Disabling resolution of transitive dependencies
- New content was added to "Declaring dependencies" to translate existing, valuable content.
    - Resolving specific artifacts from a module dependency
- The existing "Optional attributes" was deleted for now. We may want to think about just adding a sample as code and referring to to it somewhere from the text.
- The section "Working with dependencies" doesn't quite fit into this section so I left it as-is. I can see this content being covered in "Inspecting dependencies" to show how to inspect the resolution result.

jjohannes commented:
Thanks @bmuschko. I will have a look at this on Monday.

bmuschko commented:
Thanks, @jjohannes. Take your time. Beginning of next week is good enough. It's not going to be shipped with 4.5.

bmuschko commented:
@eriwen @jjohannes Would be great to get a review for this on Monday already if possible. The content has been shuffled around quite a bit so every follow-up change would likely have a merge conflict that's hard to resolve. Thanks a lot!

jjohannes commented on subprojects/docs/src/samples/userguide/dependencies/unresolvedTransitiveDependencies/build.gradle:
```diff
@@ -0,0 +1,11 @@
+//START SNIPPET unresolved-transitive-dependencies
+apply plugin: 'java-library'
```
Nitpick: `apply plugin: 'java'` instead

Otherwise it would be an example with bad practice. A library should never depend on a logging framework implementation. Only on it's API.

bmuschko replied:
You are right. Good comment. So we'd use the `java` plugin + the `compile` configuration here right?

bmuschko replied:
Changed it accordingly.

jjohannes replied:
No it should still use `implementation`. That is also available in the `java` plugin. basically `implementation` is the _new_ compile. But we did not call it compile, to not break the existing compile. 

I noticed however, that we did not make the change in the `java` plugin docs. We need to adjust it there as well. I added a comment here: https://github.com/gradle/dotorg-docs/issues/42#issuecomment-356232878

bmuschko replied:
done

bmuschko commented on subprojects/docs/src/docs/userguide/dependencyManagement.adoc:
```diff
@@ -198,102 +198,149 @@ Let's say you wanted to declare a dependency on the link:https://tomcat.apache.o
-Should you exclude per-dependency or per-configuration? It turns out that in the majority of cases you want to use the per-configuration exclusion. Here are some typical reasons why one might want to exclude a transitive dependency. Bear in mind that for some of these use cases there are better solutions than exclusions!
+[NOTE]
+====
+As a build script author you often times know that you want to exclude a dependency for all configurations available in the project. The Groovy DSL uses meta programming capabilities to allow the use of the `all` keyword to define a global rule e.g. `configurations.all { ... }`.
```
I specifically mentioned Groovy here because I am not sure if we have the same handling for Kotlin. @bamboo @eskatos Is this supported by the Kotlin DSL and if yes how does it look like?

eskatos replied:
`configurations.all {}` is no magic, see [DomainObjectCollection#all](https://github.com/gradle/gradle/blob/b0032a8a355379e5868b22d660f8be5cf857168e/subprojects/core-api/src/main/java/org/gradle/api/DomainObjectCollection.java#L126).
It is just a regular method invocation on both the Groovy and Kotlin DSLs.

bmuschko replied:
Thanks for clarifying. For some reason I thought that it was based on metaprogramming. Will change the text here.

bmuschko commented on subprojects/docs/src/docs/userguide/dependencyManagement.adoc:
```diff
@@ -198,102 +198,149 @@ Let's say you wanted to declare a dependency on the link:https://tomcat.apache.o
 
-There are other situations where it is necessary to fine-tune dependency declarations.
-Please see the api:org.gradle.api.artifacts.dsl.DependencyHandler[] class in the API documentation for examples and a complete reference for declaring dependencies.
+link:https://developers.google.com/speed/libraries/[Google Hosted Libraries] is a distribution platform for popular, open-source JavaScript libraries. With the help of the artifact-only notation you can download a JavaScript library file e.g. JQuery. The `@` character separates the dependency's coordinates from the artifact's file extension. The api:org.gradle.api.artifacts.dsl.DependencyHandler[ ...
```
OK, I will remove it for now.

jjohannes replied:
Since this documentation mainly targets new users in my mind, I think removing this here is fine. There are a lot of details and additional functionality which is not documented in this section. If I am a more advanced user and already know about map notation, I can still find information on it (e.g. in the API docs of DependencyHandler).

bmuschko commented on subprojects/docs/src/docs/userguide/dependencyManagement.adoc:
````diff
@@ -198,102 +198,149 @@ Let's say you wanted to declare a dependency on the link:https://tomcat.apache.o
+      https://repo.maven.apache.org/maven2/com/sun/jmx/jmxri/1.2.1/jmxri-1.2.1.jar
+```
+
+The situation can be fixed by excluding the dependencies that do not exist in the declared repositories for the particular dependency declaration. Exclusions need to spelled out as a key/value pair via the attributes `group` and/or `module`. For more information, refer to api:org.gradle.api.artifacts.ModuleDependency#exclude(java.util.Map)[].
````
Yeah, it already took me quite some time to come up with the example I have right now.

> I am not sure what exactly happens in this case. Maybe it's a good example if log4j really runs without the dependencies (and does not throw a runtime error later).

I will look into it. Maybe the metadata is just wrong and the dependencies are unnecessary.

bmuschko replied:
Dug deeper. The JMS and JMX dependencies are only needed if you use the relevant Log4J functionality a JMS-based appender. Changed the reasoning for the exclusion.

bmuschko commented:
@jjohannes 

> I added some suggestions, but I am fine if you merge the current state and, possibly, address some comments later.

Addressed all your comments. Could you have another look?

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagement.adoc:
```diff
@@ -198,102 +198,151 @@ Let's say you wanted to declare a dependency on the link:https://tomcat.apache.o
+++++
 
-Basically, in most of the cases excluding the transitive dependency should be done per configuration. This way the dependency declaration is more explicit. It is also more accurate because a per-dependency exclude rule does not guarantee the given transitive dependency does not show up in the configuration. For example, some other dependency, which does not have any exclude rules, might pull in t ...
+If the project requires a specific version of a dependency on a configuration-level then it can be achieved by calling the method api:org.gradle.api.artifacts.ResolutionStrategy#force(java.lang.Object...)[]. Forced dependencies on the configuration-level only work if the configuration is resolvable; that is api:org.gradle.api.artifacts.Configuration#isCanBeResolved()[] returns `true`.
```
"if the configuration is resolvable" — I don't know what this means. In what user-facing scenario would a configuration not be "resolvable"?

bmuschko replied:
This refers to the configurations defined and described in the [Java Libray plugin](https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_configurations_graph). I'll be honest I am not sure what the exact design decision was here. The documentation doesn't really explain it. The same is true for "consumable".

@jjohannes @melix Could you please clarify?

jjohannes replied:
Thanks for bringing this up @eriwen. I also stumbled over this in my review. But I did not have a good idea how to improve it then. 

A general problem I think is that too much "implementation details" about _configurations_ bleed into the API atm. This, again, has historical reasons. _Configuration_ became such a central concept in the implementation, that it is responsible for "everything" in dependency resoultion.
(The long term fix for this will be to turn the whole configuration concept more and more into an internal detail and providing new user facing APIs that talk about _variants_ instead of _configurations_. I think we will eventually get there in a year or so.)

For now, we have two kinds of configurations
1. Configurations to declare dependencies (these are now usually **not** _resolvable_)
2. Configurations to consume (these are _resolvable_)

This is, because configurations can be combined (via _extends_). So, for example, we have a couple of configs to declare dependencies that end up in the compile classpath. But we have only one _compileClasspath_ configuration that combines them all.
Because in the end, a consumer only cares about the complete classpath. And not the details, where exactly each dependency was declared.

So what can we do here? Maybe add an explanation at the place where configurations are introduced. Paraphrasing what I wrote above. Something along the lines of:
_There are two kind of configurations. One to declare dependencies and one that is exposed to tasks in the build to consume dependencies. For example, dependencies you require to compile a java project are defined using the `implementation` configuration. The compile task then uses the `compileClasspath` configuration which combines the dependencies defined in `implementation` with all other dependencies needed to compile the project (e.g. transitive dependencies)_

I have mixed feeling about investing too much effort to explaining this better. Because eventually I think the API should be improved by leveraging variant aware dependency management. But that will not happen soon. So it's probably worth documenting this better now.

eriwen replied:
Thank you for the explanation @jjohannes. So concretely: `compile` is not "resolvable" but .. wait now can you clarify exactly what you mean by "Configurations to consume"? A concrete example would be the very best way to explain here, I think.

Folks are going to have to deal with this for many years until they can upgrade to 6.0 or later. We _must_ document this. Clearly. More documentation is the price we pay for developing complex concepts.

bmuschko replied:
Thanks for the details, @jjohannes. I am going to remove the comments here for now. The meaning of the terms should rather be documented in a new section that explains it more generally. The implications also need to be explained in the docs on the `java` and `java-library` plugin. I will work on this soon.

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagement.adoc:
```diff
@@ -198,102 +198,151 @@ Let's say you wanted to declare a dependency on the link:https://tomcat.apache.o
+Disabling transitive dependency resolution will likely require you to declare the necessary runtime dependencies in your build script which otherwise would have been resolved automatically. Not doing so might lead to runtime classpath issues.
+====
+
+A project can decide to do away without transitive dependency resolution completely. You either don't want to rely on the metadata published to the consumed repositories or you want to gain full control over the dependencies in your graph. Transitive dependency resolution can be disabled on a configuration-level for any configuration returning `true` for api:org.gradle.api.artifacts.Configuration ...
```
"do away without" is a bit off to my American mind. "Disable" is [Strunk and White](https://en.wikipedia.org/wiki/The_Elements_of_Style) IMO

eriwen replied:
"returning `true` for ..." probably can be shortened to "resolvable" once we clarify what that is above.

bmuschko replied:
> "do away without" is a bit off to my American mind. "Disable" is Strunk and White IMO

done

eriwen commented on subprojects/docs/src/samples/userguide/dependencies/disablingTransitiveDependencyResolutionForConfiguration/build.gradle:
```diff
@@ -0,0 +1,22 @@
+
+//START SNIPPET transitive-per-configuration
+configurations {
+    all {
```
Better to write `configurations.all {}` maybe to clarify that `all` is a method and not a configuration name?

bmuschko replied:
It's another case of being able to achieve the same thing in different ways. We should strive to reduce it as much as we can. It's simply confusing users. I will change it as proposed.

eriwen commented on subprojects/docs/src/samples/userguide/dependencies/disablingTransitiveDependencyResolution/settings.gradle:
```diff
@@ -0,0 +1 @@
+rootProject.name = 'deactivating-transitive-dependency-resolution'
```
Is it an issue that the project name doesn't really match the directory? "deactivating" vs. "disabling"

bmuschko replied:
It's not an issue. It's more what the subdirectory was called in the process when I developed it. I changed it to "disabling".

eriwen replied:
👍

jjohannes commented:
@bmuschko good work! Would be nice if you can change the remaining `compile` to `implementation`. See my comment above. 

I think the whole _resolvable configuration_ issue should be addressed separately. Maybe in the context of reworking the _java plugin_ section. Because that's the place where understanding it is really needed. (For simple custom configurations you do not need to care about that.)

eriwen commented:
@bmuschko we're working on Java plugin docs next, so if you want to address "resolvable" configuration bits, then that's fine and we'll circle back to linking from this doc very soon.
